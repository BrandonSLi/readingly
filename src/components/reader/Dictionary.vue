<template>
    <modal 
        classes="dictionary" 
        name="dictionary" 
        :adaptive="true" 
        :min-height="200"
    >
        <div class="dictionary__word">{{ word }}</div>

        <div v-if="loading">Loading...</div>
        <div v-else-if="isEmpty(translations)">We cannot find translations of this word</div>
        <div class="dictionary__translations" v-else>
            <div 
                v-for="(words, key) in translations"
                :key="key"    
            >   
                <div class="dictionary__key">{{ key }}</div>
                <div>{{ words.join(' ‚óè ') }}</div>
            </div>
        </div>
    </modal>
</template>

<script>
import { get } from 'axios'
import { isEmpty } from 'lodash'

export default {
    name: 'Dictionary',
    props: {
        word: {
            type: String,
            required: true
        }
    },
    data() {
        return {
            translations: {},
            loading: false
        }
    },
    mounted() {
        if (! this.word) return

        this.fetchTranslations()    
    },
    methods: { 
        isEmpty,
        async fetchTranslations() {
            this.loading = true
            const { data } = await get(process.env.VUE_APP_NOW_API_URL + `/translate/${ this.word }`)
            this.loading = false

            this.translations = data
        }
    },
    watch: {
        word() {
            this.fetchTranslations()
        }
    }
}
</script>

<style lang="scss">
    .dictionary {
        background: #fff;
        padding: 1em;
        overflow: auto !important;

        &__word {
            font: {
                weight: 600;
                size: 1.2em;
            }
            color: #ff7315;
            text-transform: uppercase;
            padding-bottom: 1em;
        }

        &__key {
            font-weight: 600;
        }

        &__translations > :not(:first-child) {
            padding-top: 1em;
        }

        & * {
            text-align: center;
        }
    }
</style>