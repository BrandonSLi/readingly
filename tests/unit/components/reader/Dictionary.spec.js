import { shallowMount } from '@vue/test-utils'
import Dictionary from '@/components/reader/Dictionary'
import moxios from 'moxios'
import Vue from 'vue'
Vue.config.ignoredElements = ['modal']

describe('Dictionary', () => {
    beforeEach(() => moxios.install())
    afterAll(() => moxios.uninstall())

    it('shows info when translations are not available', async (done) => {
        moxios.stubRequest(process.env.VUE_APP_NOW_API_URL + '/translate/badword', {
            response: {}
        })

        const wrapper = shallowMount(Dictionary, {
            propsData: {
                word: 'badword'
            }
        })

        await wrapper.vm.$nextTick()
        expect(wrapper.html()).toContain('Loading...')

        moxios.wait(() => {
            expect(wrapper.html()).toContain('We cannot find translations of this word')
            done()
        })
    })

    it('shows translations when are available', (done) => {
        moxios.stubRequest(process.env.VUE_APP_NOW_API_URL + '/translate/mutter', {
            response: {
                'mutter {noun}': [ 'mamrotanie' ],
                'to mutter {vb}': [ 'szemrać', 'mamrotać', 'przebąkiwać', 'mruknąć' ]
            }
        })

        const wrapper = shallowMount(Dictionary, {
            propsData: {
                word: 'mutter'
            }
        })

        expect(wrapper.html()).toContain('Loading...')

        moxios.wait(() => {
            expect(wrapper.html()).toMatchSnapshot()
            done()
        })
    })
})