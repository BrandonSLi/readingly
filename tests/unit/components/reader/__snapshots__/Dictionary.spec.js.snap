// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dictionary shows translations when are available 1`] = `
<modal classes="dictionary" name="dictionary" adaptive="true" min-height="200">
  <div class="dictionary__word">mutter</div>
  <div class="dictionary__translations">
    <div>
      <div class="dictionary__key">mutter {noun}</div>
      <div>mamrotanie</div>
    </div>
    <div>
      <div class="dictionary__key">to mutter {vb}</div>
      <div>szemrać ● mamrotać ● przebąkiwać ● mruknąć</div>
    </div>
  </div>
</modal>
`;
